contracts/staking/utils/StakingRewardsPerEpoch.sol

```diff
@@ -0,0 +1,166 @@
+			return rewardsPerTokenAt[lastEpochIndex] - rewardsPerTokenAt[userEpochIndex];
+		} else {
+			return 0;
+		}
+	}
+
+	/**
+	 * @dev The function allows anyone to calculate the exact amount of reward
+	 * earned per epochs passed.
+	 * @param account A staker address
+	 */
+	function earned(address account) public view returns (uint256) {
+		return
+			(stakersInfo[account].balance * _getRewardPerTokenPerUser(account))
+				/ PRECISION
+				+ stakersInfo[account].reward;
+	}
+
+	function _addPendingStakesToBalanceOnEpoch(address _account) internal {
+		// If stakers balance weren't updated when he staked and the staked
+		// amount is greater than 0, then update active earning balance, nullify
+		// pending one and update global sum of pending stakes.
+		if (stakersInfo[_account].indexOfLastEpochStaked != lastEpochIndex
+				&& stakersInfo[_account].pendingStake > 0) {
+			stakersInfo[_account].balance += stakersInfo[_account].pendingStake;
+			stakersInfo[_account].pendingStake = 0;
+			pendingStakes -= stakersInfo[_account].pendingStake;
+		}
+	}
+
+	// this function updates the reward for the specific user
+	function _updateReward(address _account) internal virtual {
+		_addPendingStakesToBalanceOnEpoch(_account);
+		stakersInfo[_account].reward = earned(_account);
+	}
+
+	// This function adds the sum given in reward parameter to the distribution
+	// queue.
+	function _notifyRewardAmount(uint256 reward) internal {
+		// update cumulative rewards
+		rewardsPerTokenAt.push(
+			rewardsPerTokenAt[rewardsPerTokenAt.length - 1]
+				+ (reward * PRECISION) / totalSupply
+		);
+		// turn pending stakes to active stakes
+		totalSupply += pendingStakes;
+		pendingStakes = 0;
+
+		// update epoch count
+		lastEpochIndex++;
+	}
+
+	function _withdraw(address _from, uint256 _amount) internal virtual {
+		// if there are any pending stake for _from
+		if (stakersInfo[_from].pendingStake > 0) {
+			// if requested sum to withdraw is higher than actual pending stake balance
+			// then just withdraw the current balance
+			uint256 pendingToReduce = stakersInfo[_from].pendingStake >= _amount
+				? _amount
+				: stakersInfo[_from].pendingStake;
+			pendingStakes -= pendingToReduce;
+			stakersInfo[_from].pendingStake -= pendingToReduce;
+			stakersInfo[_from].balance -= _amount - pendingToReduce;
+		} else {
+			// elsewise just withdraw the active balance
+			stakersInfo[_from].balance -= _amount;
+		}
+		emit Withdrawn(_from, _amount, lastEpochIndex);
+	}
+
+	function _stake(address _from, uint256 _amount)
+  	internal
+  	virtual
+  {
+		// the _from address could stake to the pending balance an above zero sum
+		require(_amount > 0, "Cannot stake 0");
+		pendingStakes += _amount;
+		stakersInfo[_from].pendingStake += _amount;
+		stakersInfo[_from].indexOfLastEpochStaked = lastEpochIndex;
+		emit Staked(_from, _amount, lastEpochIndex);
+	}
+
+	function _getReward(address _to) internal virtual returns(uint256 reward) {
```

Comment:
This is internal, do we call it anywhere? Reward is subtracted from user but not sent... we better discuss it